front-end/src/main/kotlin/org/jetbrains/kotlin/spec/entity/SpecEntities.kt

```diff
@@ -0,0 +1,68 @@
+package org.jetbrains.kotlin.spec.entity
+
+import org.jetbrains.kotlin.spec.entity.test.Test
+import org.jetbrains.kotlin.spec.entity.test.parameters.LinkType
+import org.jetbrains.kotlin.spec.entity.test.parameters.TestType
+import org.jetbrains.kotlin.spec.entity.test.parameters.testArea.TestArea
+
+private interface SpecEntity {
+    val loadedTestAreas: MutableMap<TestArea, List<Test>>
+}
+
+class Section(override val loadedTestAreas: MutableMap<TestArea, List<Test>>) : SpecEntity
+
+class Paragraph(section: Section, val paragraph: Int) : SpecEntity {
+    override val loadedTestAreas: MutableMap<TestArea, List<Test>> = mutableMapOf()
+
+    init {
+        section.loadedTestAreas.forEach { map ->
+            loadedTestAreas[map.key] = map.value.filter { test -> test.testPlace.paragraph == paragraph }
+        }
+    }
+}
+
+class Sentence(paragraph: Paragraph, sentence: Int) : SpecEntity {
+    override val loadedTestAreas: MutableMap<TestArea, List<Test>> = mutableMapOf()
+
+    init {
+        paragraph.loadedTestAreas.forEach { map ->
+            loadedTestAreas[map.key] = map.value.filter { test -> test.testPlace.sentence == sentence }
+        }
+    }
+
+    fun getTestsByTestArea(testArea: TestArea): List<Test>? {
+        loadedTestAreas.forEach { map ->
+            if (map.key == testArea)
+                return if (map.value.isEmpty())
+                    null else map.value
+        }
+        return null
+    }
+
+    fun getTestsByTestType(testArea: TestArea, testType: TestType): List<Test>? {
+        val result = getTestsByTestArea(testArea)?.filter { test -> test.testPlace.testType == testType }
+        return when {
+            result == null -> null
+            result.isEmpty() -> null
+            else -> result
+        }
+    }
+
+    fun getTestsByTestPriority(testArea: TestArea, testType: TestType, linkType: LinkType): List<Test>? {
+        val result = getTestsByTestType(testArea, testType)?.filter { test -> test.testInfo.linkType == linkType }
```

Comment:
BTW, you can unify this and use empty lists everywhere as a marker for the lack of tests.

Then you can simplify these methods and use just `orEmpty`:
```
fun getTestsByTestPriority(testArea: TestArea, testType: TestType, linkType: LinkType) =
    getTestsByTestType(testArea, testType)?.filter { test -> test.testInfo.linkType == linkType }.orEmpty()
```

(or opposite: use `null` everywhere, without `isEmpty` checks)